macrosynergy/download/dq_api.py

```diff
@@ -0,0 +1,749 @@
+    :param <str> url:
+    :param <str> token_url:
+    :param <str> dq_resource_id:
+    """
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        url: str = OAUTH_BASE_URL,
+        token_url: str = OAUTH_TOKEN_URL,
+        dq_resource_id: str = OAUTH_DQ_RESOURCE_ID,
+        token_proxy: Optional[dict] = None,
+    ):
+
+        self.base_url: str = url
+        self.__token_url: str = token_url
+        self.__dq_api_resource_id: str = dq_resource_id
+
+        id_error = f"client_id argument must be a <str> and not <{type(client_id)}>."
+        assert isinstance(client_id, str), id_error
+        self.client_id: str = client_id
+
+        secret_error = f"client_secret must be a str and not <{type(client_secret)}>."
+        assert isinstance(client_secret, str), secret_error
+
+        self.client_secret: str = client_secret
+        self._stored_token: Optional[dict] = None
+        self.token_data = {
+            "grant_type": "client_credentials",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "aud": self.__dq_api_resource_id,
+        }
+
+        # For debugging purposes save last request response.
+        self.status_code: Optional[int] = None
+        self.last_response: Optional[str] = None
+        self.last_url: Optional[str] = None
+        self.token_proxy: Optional[dict] = token_proxy
+
+    def _active_token(self) -> bool:
+        """Confirms if the token being used has not expired."""
+        created: datetime = self._stored_token["created_at"]
+        expires: int = self._stored_token["expires_in"]
+
+        return (datetime.now() - created).total_seconds() / 60 >= (expires - 1)
+
+    def _valid_token(self) -> bool:
+        """Confirms if the credentials passed correspond to a valid token."""
+        return not (self._stored_token is None or self._active_token())
+
+    def _get_token(self) -> str:
+        """Retrieves the token which is used to access DataQuery via OAuth method."""
+        i = 0
+        success = False
+        while (not success) and (i < 3):
+            js, success, self.last_url, msg = dq_request(
+                url=self.__token_url,
+                data=self.token_data,
+                method="post",
+                proxies=self.token_proxy,
+            )
+            i += 1
+
+        runtime_error = RuntimeError(f"Unable to retrieve token. Error: {msg}. ")
+        if success:
+            try:
+                self._stored_token: dict = {
+                    "created_at": datetime.now(),
+                    "access_token": js["access_token"],
+                    "expires_in": js["expires_in"],
+                }
+            except:
+                raise runtime_error
+            else:
+                if self._valid_token():
+                    return self._stored_token["access_token"]
+        else:
+            raise runtime_error
```

Comment:
Why `RuntimeError`?

Reply:
This was for the case when the process fails to get a new valid token. I had assumed it was the correct type of error to use; have reverted to the original method as of now

Reply:
```python
class AuthenticationError(RuntimeError):
    """HTTP AuthenticationError"""
```